src/c++/perf_analyzer/client_backend/openai/http_client.cc

```diff
@@ -183,94 +183,102 @@ HttpClient::SetSSLCurlOptions(CURL* curl_handle)
 void
 HttpClient::Send(CURL* handle, std::unique_ptr<HttpRequest>&& request)
 {
-  std::lock_guard<std::mutex> lock(mutex_);
-
-  auto insert_result = ongoing_async_requests_.emplace(
-      std::make_pair(reinterpret_cast<uintptr_t>(handle), std::move(request)));
-  if (!insert_result.second) {
-    curl_easy_cleanup(handle);
-    throw std::runtime_error(
-        "Failed to insert new asynchronous request context.");
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    auto insert_result = ongoing_async_requests_.emplace(std::make_pair(
+        reinterpret_cast<uintptr_t>(handle), std::move(request)));
+    if (!insert_result.second) {
+      curl_easy_cleanup(handle);
+      throw std::runtime_error(
+          "Failed to insert new asynchronous request context.");
+    }
+    curl_multi_add_handle(multi_handle_, handle);
   }
-  curl_multi_add_handle(multi_handle_, handle);
+
   cv_.notify_all();
+
+  // Wake up the async transfer thread to handle new requests immediately
+  curl_multi_wakeup(multi_handle_);
 }
 
 void
 HttpClient::AsyncTransfer()
 {
   int place_holder = 0;
   CURLMsg* msg = nullptr;
-  do {
+
+  while (!exiting_) {
     std::vector<std::unique_ptr<HttpRequest>> request_list;
 
-    // sleep if no work is available
+    // Lock the mutex to check for work and wait if necessary
     std::unique_lock<std::mutex> lock(mutex_);
     cv_.wait(lock, [this] {
-      if (this->exiting_) {
-        return true;
-      }
-      // wake up if an async request has been generated
-      return !this->ongoing_async_requests_.empty();
+      return this->exiting_ || !this->ongoing_async_requests_.empty();
     });
 
+    if (exiting_) {
+      break;
+    }
+
+    lock.unlock();  // Release the mutex before waiting on network activity
+
+    // Perform any pending transfers
     CURLMcode mc = curl_multi_perform(multi_handle_, &place_holder);
+    if (mc != CURLM_OK) {
+      std::cerr << "Unexpected error: curl_multi_perform failed. Code:" << mc
+                << std::endl;
+      continue;
+    }
+
+    // Wait for activity on the multi_handle_
     int numfds;
-    if (mc == CURLM_OK) {
-      // Wait for activity. If there are no descriptors in the multi_handle_
-      // then curl_multi_wait will return immediately
-      mc = curl_multi_wait(multi_handle_, NULL, 0, INT_MAX, &numfds);
-      if (mc == CURLM_OK) {
-        while ((msg = curl_multi_info_read(multi_handle_, &place_holder))) {
-          uintptr_t identifier = reinterpret_cast<uintptr_t>(msg->easy_handle);
-          auto itr = ongoing_async_requests_.find(identifier);
-          // This shouldn't happen
-          if (itr == ongoing_async_requests_.end()) {
-            std::cerr
-                << "Unexpected error: received completed request that is not "
-                   "in the list of asynchronous requests"
+    mc = curl_multi_poll(multi_handle_, NULL, 0, 1000, &numfds);
```

Comment:
When the other thread wakes up `curl_multi_poll` is it going to change the status code?

Reply:
Thanks for reviewing this PR! Do you mean the return value? Yes, it'll be updated. You can see the example code for it in the documentation [here](https://curl.se/libcurl/c/curl_multi_poll.html).